design/0003-service-topology-support.md

```diff
@@ -0,0 +1,111 @@
+
+Toplogy metadata that is obtained from supporting this feature allows for prioritising endpoints based on locality.
+
+(_Adapted from [#4325](https://github.com/linkerd/linkerd2/issues/4325)_)
+(_N.B: I use locality and topology interchangeably here, when used in the same context to me both mean the same thing, apologies in advance if the language barrier created any confusion._)
+
+## Design proposal (Step 2)
+
+[design-proposal]: #design-proposal
+
+### Summary
+---
+
+To implement support for this feature, most of the work will be carried out in the `destination service` on the control plane side. Although the feature does not add a lot of changes in terms of the Kubernetes API, several changes have to be made. The first and most obvious is to support the `EndpointSlices` feature that was introduced in Kubernetes `v1.16`, since this will enable endpoints to hold topology metadata.
+
+On the proxy side of all meshed pods, the outbound traffic is load balanced between all available endpoints for a targeted service. Through this change, we will introduce a new priority associated with each endpoint based on its locality. If a service has a topological preference defined in its `spec` through the newly added `topologyKeys`, then the proxy will receive only the endpoints of the highest priority that are available and satisfy the `topologyKeys` preference. In order to determine the priority of endpoints, the proxy will have to send an additional contextual token as part of its service discovery request: the `nodeName` of the host the meshed pod is deployed on. When a proxy will first resolve the FQDN or IP of the outbound service, the `Destination.Get` call will include the topology context token, along with the rest of the payload.
+
+Whenever the destination service receives a request from a proxy, it resolves the FQDN or IP of a service, creates a listener (as an `EndpointTranslator`) and creates a subscription on the service. At first, all available endpoints for the service are sent back as a `WeightedAddressSet`, all subsequent endpoint creations or deletions are propagated through the listener back to the proxy. In order to do this, the destination service makes use of watchers to cache and register any Kubernetes resource changes (such as modified endpoints, newly created endpoints or services that are deleted). Through this feature, every service that is registered by the watchers and cached would include additional `topologyKeys` labels. Likewise, endpoints would now have their topological information stored, this would be done by adding support for the `EndpointSlices` feature. By using the proxy's topology context token (the `nodeName` mentioned previously), the `EndpointTranslator`, which acts as the listener implementation, can now assign priorities to each endpoint and only send the endpoint that satisfy the highest priority required by the service and available in the cluster.
```

Comment:
> and only send the endpoint that satisfy the highest priority required by the service and available in the cluster.

Is this correct? What if none of these endpoints are available? Shouldn't the proxy be able to use other endpoints? Or what if the highest priority endpoints are extremely slow? Shouldn't the proxy be able consider lower-priority endpoints for the purposes of balancing?

Reply:
@adleong points out this from the [docs](https://kubernetes.io/docs/concepts/services-networking/service-topology/#using-service-topology):

>  Traffic will be directed to a Node whose value for the first label matches the originating Node’s value for that label. If there is no backend for the Service on a matching Node, then the second label will be considered, and so forth, until no labels remain.

Reply:
Thank you for the comment! 

> What if none of these endpoints are available? Shouldn't the proxy be able to use other endpoints?

This depends on the service's topology preference. Priority is used a bit loosely here, in the sense that nodes are prioritised to be sent if they match the preferred topological domain, if none exist then we simply fallback to the next preference in the list.

Assuming the following preference-ordered list `[hostname, topology/zone, topology/region]` for a service, then the highest priority endpoints would be the endpoints on the same node as the source. If no such endpoints exist, then the next topological domain is taken (in this case `topology/zone`) and so on until the list is exhausted. If the list is exhausted and there are no endpoints on the same node, same zone or same region as the source (the pod calling the service in question) then there would be no endpoints. 

Suppose the following topologyKeys list is used instead for the service: `[hostname]`. In this case if there are no endpoints on the same node as the source pod, then the proxy would not receive any endpoints. Additionally, the `*` can be used to indicate "any endpoints" (official docs suggest this should be used at the end of the list to provide a reliable fallback). In such a case, if an operator specifies the following preference for a service: `[hostname, *]`, then if there are no endpoints on the same node as the source pod, the proxy will receive all other endpoints of that service as a fallback.

> Or what if the highest priority endpoints are extremely slow? Shouldn't the proxy be able to consider lower-priority endpoints for the purpose of balancing?

I am very glad you brought up this point. I was initially thinking along the same lines but from the perspective of availability -- i.e what if we only have 1 available endpoint for the top-most preference, that would surely lead to congestion and overload the endpoint, it would make more sense to send a proportion of the endpoints, not just the ones with the highest priority.

With that being said, @adleong made an interesting point here, namely to stick to the service topology specification, and the more I read into service topology as a K8s feature the more I agree with it.

> The Service Topology feature resolves this by allowing the Service creator to define a policy for routing traffic based upon the Node labels for the originating and destination Nodes. (Service Topology Docs)

Service topology would allow operators, users, etc. to create their own routing policies based on locality, I believe if we do not adhere to the preferential list of the service then we will likely introduce side effects. Perhaps this feature should only enforce the policy users create but not modify it or extend it in any way, hope this makes sense.

I am more than happy however to make changes and amend the RFC as well as go in a completely different direction with this if requested. Let me know your thoughts!